### PR TITLE
Implement aten::index | feat(torchlib)

### PR DESCRIPTION
**Pull request by justinchuby**
WIP: need more tests

Gather op:
- https://github.com/openxla/xla/blob/main/docs/operation_semantics.md?rgh-link-date=2023-07-13T01%3A09%3A16Z#gather
- https://www.pathpartnertech.com/gather-scatter-operation-in-deep-learning-framework/

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -3026,10 +3028,54 @@ def aten_imag(self: TensorType) -> TensorType:
...
+        return sorted(lst) == list(range(min(lst), max(lst) + 1))
+
+
+def _is_none_in_middle(indices):
+    not_none_indices = [i for i, idx in enumerate(indices) if idx is not None]
+    return not are_consecutive(not_none_indices)
+
+
+@torch_op("aten::index", trace_only=True)
+def aten_index(self: TensorType, indices: Sequence[Optional[INT64]]) -> TensorType:
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/ops/core.py:3044]: This should work for all cases except for bool mask index, if possible.

Let me know if you can find a bug! @justinchuby

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 2]
```diff
@@ -3026,10 +3028,54 @@ def aten_imag(self: TensorType) -> TensorType:
...
 
-    raise NotImplementedError()
+    ordered_indices = sorted(range(len(indices)), key=lambda i: (indices[i] is None, i))
+    ordered_indices += list(range(len(ordered_indices), len(self.shape)))
+
+    self = op.Transpose(self, perm=ordered_indices)
+
+    # Need broadcast concat lol.
+    not_none_indices = [idx for idx in indices if idx is not None]
+    broadcast_shape = np.broadcast_shapes(*[idx.shape for idx in not_none_indices])
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:3054]: We may not have shapes here: need to assert or use alternatives like implicit broadcasting?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 3]
```diff
@@ -3026,10 +3028,54 @@ def aten_imag(self: TensorType) -> TensorType:
...
+
+    self = op.GatherND(self, final_index, batch_dims=0)
+    if _is_none_in_middle(indices):
+        return self
+
+    # Need to transpose back.
+    adv_res_rank = len(broadcast_shape)
+    # [adv1, adv2, ..., adv_res_rank, x1, x2, ..., xk, ..., xn] -> [x1, x2, ..., xk, adv1, ..., adv_res_rank, ..., xn]
+    perm = (
+        list(range(adv_res_rank, adv_res_rank + ordered_indices[0]))
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:3068]: I would unpack instead

File: onnxscript/tests/function_libs/torch_lib/extra_opinfo.py [hunk 4]
```diff
@@ -446,6 +446,22 @@ def sample_inputs_col2im(op_info, device, dtype, requires_grad, **kwargs):
...
+def sample_inputs_index(op_info, device, dtype, requires_grad, **kwargs):
+    del op_info  # Unused
+    del kwargs  # Unused
+    make_arg = functools.partial(
+        torch_testing.make_tensor, dtype=dtype, device=device, requires_grad=requires_grad
+    )
+    s = 5
+    test_args = [
+        ([common_methods_invocations.index_variable(2, s, device=device)],),
+        # ([torch.tensor()],)
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:458]: @justinchuby will add more tests

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 2 above)
Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/ops/core.py:3054]: yea that's kinda rough. If we don't, or if there are symbols, we need to dynamically broadcast, probably via `add`. However at some point later we need the rank of broadcasted tensor, which is needed to construct the `perm` for transpose. If we cannot get it statically, then the transpose becomes a headache.

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 3 above)
Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/ops/core.py:3068]: feel free to make changes. the current version is quick hack style

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:3054]: I think we can assume the rank. This is more relaxed than assuming shape

File: onnxscript/tests/function_libs/torch_lib/extra_opinfo.py (hunk 4 above)
Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:458]: Sharing snippet I used to test. It poses complication here as indexing, i.e., `data[]`, lowers to multiple operators, usually couple of `aten.slice` + `aten.index` in the end. So you might need to figure out a way to construct the proper input state if you are testing `aten.index.Tensor` alone.

```python
import torch
import onnxruntime
import onnxscript


class IndexModel(torch.nn.Module):
    def forward(self, data, index, index2):
        # return data[..., index, index2]
        # return data[index, :, index2, ...]
        # return data[..., index, index2, :]
        # return data[index, :, index2]
        return data[:, :, index, index2]


data = torch.arange(0, 7 * 3 * 4 * 5 * 6).view(7, 3, 4, 5, 6)
index = torch.tensor([2, 1])
index2 = torch.tensor([[[0]], [[2]]])

model = IndexModel()
pt_output = model(data, index, index2)
print("PyTorch output shape:", pt_output.shape)

export_output = torch.onnx.dynamo_export(model, data, index, index2)
export_output.diagnostic_context.dump("log.sarif")

print(onnxscript.proto2text(export_output.model_proto))
sess = onnxruntime.InferenceSession(
    export_output.model_proto.SerializeToString(), providers=["CPUExecutionProvider"]
)
output = sess.run(None, {"arg0": data.numpy(), "arg1": index.numpy(), "arg2": index2.numpy()})

torch.testing.assert_close(pt_output.numpy(), output[0])

print("PASS")
```

**Comment by justinchuby** [comment]
Pictures: https://github.com/openxla/xla/blob/main/docs/operation_semantics.md#informal-description-and-examples

**Comment by gramalingam** [comment]
Is this migrating the logic used by the previous exporter (for this op)? Or, is this different or new?

**Comment by BowenBao** [comment]
> Is this migrating the logic used by the previous exporter (for this op)? Or, is this different or new?

It is new and hopefully more robust, while the logic is similar.

The old implementation was using opset9.GatherElements, at that point GatherND does not exist yet.

**Comment by gramalingam** [comment]
> The old implementation was using opset9.GatherElements, at that point GatherND does not exist yet.

Oh ... I thought Spandan introduced the new Gather op variants to support the pytorch converter.

**Comment by BowenBao** [comment]
> Oh ... I thought Spandan introduced the new Gather op variants to support the pytorch converter.

Yes it was used in some other ops, as well as extended edge case support for `aten.index` after opset11, but there wasn't a full rewrite.

**Comment by justinchuby** [comment]
Squashing to another branch
